tools/clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -3983,8 +3983,27 @@ bool Sema::CheckTemplateArgumentList(TemplateDecl *Template,
 
   // If we have any leftover arguments, then there were too many arguments.
   // Complain and fail.
-  if (ArgIdx < NumArgs)
-    return diagnoseArityMismatch(*this, Template, TemplateLoc, NewArgs);
+  // HLSL Change Starts: the SPIR-V backend supports templates with unchecked
+  // arguments
+  if (ArgIdx < NumArgs) {
+    // SPIRV change starts
+#ifdef ENABLE_SPIRV_CODEGEN
+    // Ignore arity mismatch for vk::SpirvType and vk::SpirvOpaqueType
+    // TODO(6246): this breaks if the frontend tries to desugar the type
+    NamedDecl *TemplatedDecl = Template->getTemplatedDecl();
+    llvm::StringRef TemplateName = TemplatedDecl->getName();
+    llvm::StringRef Namespace = "";
+    if (const auto *ND =
+            dyn_cast<NamespaceDecl>(TemplatedDecl->getDeclContext()))
+      Namespace = ND->getName();
+
+    if (Namespace != "vk" ||
+        (TemplateName != "SpirvType" && TemplateName != "SpirvOpaqueType"))
```

Comment:
I'm a little nervous about this. Why is this needed? Can't we just have default values for the additional template arguments?

Reply:
There's currently no way to represent `vk::SpirvType` in HLSL syntax or in the AST. The closest representation would probably be something like

```hlsl
template <uint OpCode, uint size, uint alignment, typename... args>
struct SpirvType {}
```

However, that would only allow us to pass types to `SpirvType`, whereas we would like to be able to pass either types or values of any type. So instead this PR defines `SpirvType` as

```hlsl
template <uint OpCode, uint size, uint alignment>
struct SpirvType {}
```

and modifies Sema to ignore any additional arguments. This actually doesn't work for every situation where someone might reasonably want to use `SpirvType`, so I've created #6246 as a follow-up issue.

The options to fix it will likely be either to add `typename... args` to the `SpirvType` template and specifically disable validation for that parameter pack, or to add a new type of parameter pack which will allow `SpirvType` to be accurately represented in the AST.